keanu-project/src/main/java/io/improbable/keanu/util/DescriptionCreator.java

```diff
@@ -0,0 +1,178 @@
+import static io.improbable.keanu.util.DescriptionUtils.createBooleanBinaryOpDescription;
+import static io.improbable.keanu.util.DescriptionUtils.createIfStringDescription;
+
+public class DescriptionCreator {
+
+    private static Map<Class, String> delimiters = new HashMap<>();
+    private static Map<Class, String> booleanBinaryOperators = new HashMap<>();
+
+    static {
+        delimiters.put(AdditionVertex.class, " + ");
+        delimiters.put(IntegerAdditionVertex.class, " + ");
+        delimiters.put(DifferenceVertex.class, " - ");
+        delimiters.put(IntegerDifferenceVertex.class, " - ");
+        delimiters.put(MultiplicationVertex.class, " * ");
+        delimiters.put(IntegerMultiplicationVertex.class, " * ");
+        delimiters.put(DivisionVertex.class, " / ");
+        booleanBinaryOperators.put(AndBinaryVertex.class, " && ");
+        booleanBinaryOperators.put(EqualsVertex.class, " == ");
+        booleanBinaryOperators.put(GreaterThanOrEqualVertex.class, " >= ");
+        booleanBinaryOperators.put(GreaterThanVertex.class, " > ");
+        booleanBinaryOperators.put(LessThanOrEqualVertex.class, " <= ");
+        booleanBinaryOperators.put(LessThanVertex.class, " < ");
+        booleanBinaryOperators.put(NotEqualsVertex.class, " != ");
+        booleanBinaryOperators.put(OrBinaryVertex.class, " || ");
+    }
+
+    private static <T extends Tensor> String getBaseDescription(Vertex<T> vertex) {
+        if (vertex instanceof ConstantVertex) {
+            Tensor value = vertex.getValue();
+            String valueString = value.isScalar() ? value.scalar().toString() :
+                new StringBuilder("(Tensor with shape: ").append(Arrays.toString(value.getShape())).append(")").toString();
+
+            return new StringBuilder()
+                .append("Const(")
+                .append(valueString)
+                .append(")")
+                .toString();
+        } else {
+            return "<<Unlabelled leaf Vertex>>";
+        }
+    }
+
+    static <T extends Tensor> String createDescriptionAllowingLabels(Vertex<T> vertex) {
+        Collection<Vertex> parents = vertex.getParents();
+
+        if (vertex.getLabel() != null) {
+            return vertex.getLabel().toString();
+        }
+
+        if (parents.size() == 0) {
+            return getBaseDescription(vertex);
+        }
+
+        return recursiveDescriptionStep(vertex, true);
+    }
+
+    private static <T extends Tensor> String recursiveDescriptionStep(Vertex<T> vertex, boolean includeBrackets) {
+        String irregularDescription = checkForIrregularExpressions(vertex, includeBrackets);
+        if (irregularDescription != null) {
+            return irregularDescription;
+        }
+
+        Stream<String> parentStream = vertex
+            .getParents()
+            .stream()
+            .map(DescriptionCreator::createDescriptionAllowingLabels);
+
+        String[] parentStrings = parentStream.toArray(String[]::new);
```

Comment:
Is there a reason you do this in two steps?

Reply:
Yes - for some stupid reason if I inline this then I get a compile error. Don't ask me why because I have no clue. More than happy to show you on my machine if you're interested ;)

Reply:
Ugh - I've hit similar things before, it's a pain!